File: projects/rpn2tex/migrations/go-module-by-module-1/lexer.go
package main

import (
	"fmt"
	"unicode"
)

// Lexer tokenizes RPN expression input.
// It tracks position (line and column) for error reporting.
type Lexer struct {
	Text   string
	Pos    int // Current position in Text (0-based)
	Line   int // Current line number (1-based)
	Column int // Current column number (1-based)
}

// NewLexer creates a new Lexer for the given input text.
func NewLexer(text string) *Lexer {
	return &Lexer{
		Text:   text,
		Pos:    0,
		Line:   1,
		Column: 1,
	}
}

// Tokenize scans the entire input and returns a slice of tokens.
// It returns an error if an invalid character is encountered.
func (l *Lexer) Tokenize() ([]Token, error) {
	var tokens []Token

	for !l.atEnd() {
		l.skipWhitespace()
		if l.atEnd() {
			break
		}

		token, err := l.scanToken()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, token)
	}

	// Add EOF token
	tokens = append(tokens, Token{
		Type:   EOF,
		Value:  "",
		Line:   l.Line,
		Column: l.Column,
	})

	return tokens, nil
}

// atEnd returns true if we've reached the end of input.
func (l *Lexer) atEnd() bool {
	return l.Pos >= len(l.Text)
}

// peek returns the current character without consuming it.
// Returns 0 if at end of input.
func (l *Lexer) peek() rune {
	if l.atEnd() {
		return 0
	}
	return rune(l.Text[l.Pos])
}

// peekNext returns the next character without consuming it.
// Returns 0 if at end or only one character remaining.
func (l *Lexer) peekNext() rune {
	if l.Pos+1 >= len(l.Text) {
		return 0
	}
	return rune(l.Text[l.Pos+1])
}

// advance consumes and returns the current character, updating position.
func (l *Lexer) advance() rune {
	if l.atEnd() {
		return 0
	}
	ch := rune(l.Text[l.Pos])
	l.Pos++
	if ch == '\n' {
		l.Line++
		l.Column = 1
	} else {
		l.Column++
	}
	return ch
}

// skipWhitespace consumes all whitespace characters.
func (l *Lexer) skipWhitespace() {
	for !l.atEnd() && unicode.IsSpace(l.peek()) {
		l.advance()
	}
}

// scanToken scans and returns the next token.
func (l *Lexer) scanToken() (Token, error) {
	startLine := l.Line
	startColumn := l.Column
	ch := l.advance()

	switch ch {
	case '+':
		return Token{Type: PLUS, Value: "+", Line: startLine, Column: startColumn}, nil
	case '*':
		return Token{Type: MULT, Value: "*", Line: startLine, Column: startColumn}, nil
	case '/':
		return Token{Type: DIV, Value: "/", Line: startLine, Column: startColumn}, nil
	case '-':
		// Check if this is a negative number or subtraction operator
		// Negative number: '-' followed immediately by a digit
		next := l.peek()
		if unicode.IsDigit(next) {
			return l.scanNumber("-", startLine, startColumn), nil
		}
		return Token{Type: MINUS, Value: "-", Line: startLine, Column: startColumn}, nil
	default:
		if unicode.IsDigit(ch) {
			return l.scanNumber(string(ch), startLine, startColumn), nil
		}
		// Invalid character
		return Token{}, &SyntaxError{
			Message: fmt.Sprintf("Unexpected character '%c'", ch),
			Line:    startLine,
			Column:  startColumn,
		}
	}
}

// scanNumber scans a numeric literal (integer or decimal).
// prefix is the initial character(s) already consumed (e.g., "-" or a digit).
func (l *Lexer) scanNumber(prefix string, startLine, startColumn int) Token {
	value := prefix

	// Scan integer part
	for !l.atEnd() && unicode.IsDigit(l.peek()) {
		value += string(l.advance())
	}

	// Check for decimal point
	if !l.atEnd() && l.peek() == '.' && !l.atEnd() && l.Pos+1 < len(l.Text) && unicode.IsDigit(l.peekNext()) {
		// Consume decimal point
		value += string(l.advance())
		// Scan fractional part
		for !l.atEnd() && unicode.IsDigit(l.peek()) {
			value += string(l.advance())
		}
	}

	return Token{
		Type:   NUMBER,
		Value:  value,
		Line:   startLine,
		Column: startColumn,
	}
}

File: projects/rpn2tex/migrations/java-module-by-module-2/src/main/java/com/rpn2tex/Lexer.java
package com.rpn2tex;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Tokenizes RPN expression input into a stream of tokens.
 *
 * <p>The lexer performs character-by-character scanning and recognizes:
 * <ul>
 *   <li>Numbers (integers and decimals, including negative)</li>
 *   <li>Operators: +, -, *, /</li>
 *   <li>Whitespace (used as delimiter, skipped)</li>
 * </ul>
 *
 * <p>Position tracking (line/column) is maintained for error reporting.
 * Both line and column numbers are 1-based (human-readable).
 *
 * <p>Example usage:
 * <pre>{@code
 * Lexer lexer = new Lexer("5 3 +");
 * List<Token> tokens = lexer.tokenize();
 * // tokens: [NUMBER("5"), NUMBER("3"), PLUS("+"), EOF("")]
 * }</pre>
 *
 * <p>Error handling:
 * <pre>{@code
 * try {
 *     Lexer lexer = new Lexer("2 3 ^");
 *     List<Token> tokens = lexer.tokenize();
 * } catch (RpnException e) {
 *     // e.getMessage() returns "Line 1, column 5: Unexpected character '^'"
 * }
 * }</pre>
 *
 * @since 1.0
 */
public class Lexer {
    private final String text;
    private int pos;
    private int line;
    private int column;

    /**
     * Constructs a new lexer for the given input text.
     *
     * @param text the RPN expression to tokenize
     * @throws NullPointerException if text is null
     */
    public Lexer(String text) {
        Objects.requireNonNull(text, "Input text cannot be null");
        this.text = text;
        this.pos = 0;
        this.line = 1;
        this.column = 1;
    }

    /**
     * Tokenizes the input text into a list of tokens.
     *
     * <p>The token list always ends with an EOF token.
     * Position information is captured for each token for error reporting.
     *
     * @return list of tokens ending with EOF token
     * @throws RpnException if an invalid character is encountered
     */
    public List<Token> tokenize() throws RpnException {
        List<Token> tokens = new ArrayList<>();

        while (!atEnd()) {
            skipWhitespace();
            if (atEnd()) {
                break;
            }
            tokens.add(scanToken());
        }

        tokens.add(new Token(TokenType.EOF, "", line, column));
        return tokens;
    }

    /**
     * Checks if we've reached the end of input.
     *
     * @return true if at end of input
     */
    private boolean atEnd() {
        return pos >= text.length();
    }

    /**
     * Peeks at the current character without consuming it.
     *
     * @return current character, or '\0' if at end
     */
    private char peek() {
        if (atEnd()) {
            return '\0';
        }
        return text.charAt(pos);
    }

    /**
     * Consumes and returns the current character, updating position tracking.
     *
     * <p>Position tracking:
     * <ul>
     *   <li>Newline: increments line, resets column to 1</li>
     *   <li>Other characters: increments column</li>
     * </ul>
     *
     * @return the consumed character
     */
    private char advance() {
        char c = text.charAt(pos);
        pos++;
        if (c == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        return c;
    }

    /**
     * Skips whitespace characters (space, tab, newline, carriage return).
     */
    private void skipWhitespace() {
        while (!atEnd() && Character.isWhitespace(peek())) {
            advance();
        }
    }

    /**
     * Scans and returns the next token from the input.
     *
     * <p>Token recognition:
     * <ul>
     *   <li>'+' → PLUS token</li>
     *   <li>'-' → MINUS token (or part of negative number if followed by digit)</li>
     *   <li>'*' → MULT token</li>
     *   <li>'/' → DIV token</li>
     *   <li>Digit → NUMBER token (integer or decimal)</li>
     *   <li>Other → RpnException</li>
     * </ul>
     *
     * @return the scanned token
     * @throws RpnException if an unexpected character is encountered
     */
    private Token scanToken() throws RpnException {
        int startLine = line;
        int startColumn = column;
        char c = peek();

        // Single-character operators
        if (c == '+') {
            advance();
            return new Token(TokenType.PLUS, "+", startLine, startColumn);
        }
        if (c == '-') {
            advance();
            // Check if this is a negative number (minus followed immediately by digit)
            if (!atEnd() && Character.isDigit(peek())) {
                return scanNumber("-", startLine, startColumn);
            }
            return new Token(TokenType.MINUS, "-", startLine, startColumn);
        }
        if (c == '*') {
            advance();
            return new Token(TokenType.MULT, "*", startLine, startColumn);
        }
        if (c == '/') {
            advance();
            return new Token(TokenType.DIV, "/", startLine, startColumn);
        }

        // Numbers
        if (Character.isDigit(c)) {
            return scanNumber("", startLine, startColumn);
        }

        // Unknown character
        throw new RpnException(String.format("Unexpected character '%c'", c), startLine, startColumn);
    }

    /**
     * Scans a numeric literal (integer or decimal).
     *
     * <p>Number format:
     * <ul>
     *   <li>Optional prefix (e.g., "-" for negative numbers)</li>
     *   <li>Integer part: sequence of digits</li>
     *   <li>Optional decimal part: "." followed by digits</li>
     * </ul>
     *
     * <p>Examples: 5, 3.14, -2, 0.5
     *
     * @param prefix      optional prefix (e.g., "-" for negative numbers)
     * @param startLine   line number where the token starts
     * @param startColumn column number where the token starts
     * @return NUMBER token with the numeric value as a string
     */
    private Token scanNumber(String prefix, int startLine, int startColumn) {
        StringBuilder value = new StringBuilder(prefix);

        // Integer part
        while (!atEnd() && Character.isDigit(peek())) {
            value.append(advance());
        }

        // Decimal part (optional)
        if (!atEnd() && peek() == '.') {
            value.append(advance());  // consume '.'
            while (!atEnd() && Character.isDigit(peek())) {
                value.append(advance());
            }
        }

        return new Token(TokenType.NUMBER, value.toString(), startLine, startColumn);
    }
}
